midi_tokenizers/no_loss_tokenizer.py

```diff
@@ -0,0 +1,138 @@
+import pandas as pd
+
+from midi_tokenizers.midi_tokenizer import MidiTokenizer
+
+
+class NoLossTokenizer(MidiTokenizer):
+    def __init__(
+        self,
+        eps: float = 0.001,
+    ):
+        super().__init__()
+        self.eps = eps
+        self.specials = ["<CLS>"]
+        self.vocab = self._build_vocab()
+        self.token_to_id = {token: it for it, token in enumerate(self.vocab)}
+
+    def __rich_repr__(self):
+        yield "NoLossTokenizer"
+        yield "eps", self.eps
+        yield "vocab_size", self.vocab_size
+
+    def _build_vocab(self):
+        self.vocab = list(self.specials)
+
+        # Add MIDI note and velocity tokens to the vocabulary
+        for pitch in range(21, 109):
+            self.vocab.append(f"NOTE_ON_{pitch}")
+            self.vocab.append(f"NOTE_OFF_{pitch}")
+
+        for vel in range(128):
+            self.vocab.append(f"VELOCITY_{vel}")
+
+        token = self.eps
+
+        # Generate time tokens with exponential distribution
+        while token < 1:
+            self.vocab.append(f"{token}s")
+            token *= 2
+
+        self.max_time_token = token  # Maximum time token
+        return self.vocab
+
+    @staticmethod
+    def _notes_to_event_df(notes: pd.DataFrame):
+        """
+        Convert MIDI note dataframe into a dataframe with on/off events.
+        """
+        note_on_events: pd.DataFrame = notes.loc[:, ["start", "pitch", "velocity"]]
+        note_off_events: pd.DataFrame = notes.loc[:, ["end", "pitch"]]
+
+        note_off_events["time"] = note_off_events["end"]
+        note_off_events["event"] = "NOTE_OFF"
+        note_on_events["time"] = note_on_events["start"]
+        note_on_events["event"] = "NOTE_ON"
+
+        note_events: pd.DataFrame = pd.concat([note_on_events, note_off_events], axis=0)
+        note_events = note_events.sort_values(by="time")
+
+        return note_events
```

Comment:
Looks like this tries to hard to create a dataframe that ends up being inconsistent (what's up with velocity for note off events?). I think this should return a `list[dict]` instead :thinking: (also please include return type hints always :exclamation: (unless null).

Reply:
`list[dict]` would be harder to sort, velocity for note-off events is not needed when this data is stored during note-on.

Reply:
I know velocity is not needed for note offs - that's what makes the dataframe inconsistent and that's why it converts velocity values to float, causing issues later. I think something like this would be better:

```python
note_on_events = note_on_df.to_dict(orient="records")
note_off_events = note_off_df.to_dict(orient="records")
note_events = note_on_events + note_off_events

# It's not that hard to sort a list of dicts
note_events = sorted(note_events, key=lambda event: event["time"])
return note_events
```